### ISSUE TITLE
Duplucate and non-canonically ordered map keys are accepted

### ISSUE THREAD

**Issue by DavidBuchanan314** [MarshalX/python-libipld#15]
Here are 3 test cases:

```py
import libipld

# equivalent to {"abc": 1, "abc": 2} (duplicate key)
print(libipld.decode_dag_cbor(bytes.fromhex("a263616263016361626302")))

# equivalent to {"def": 1, "abc": 2} (same key lengths, wrong sort order)
print(libipld.decode_dag_cbor(bytes.fromhex("a263646566016361626302")))

# equivalent to {"aaa": 1, "x": 2} (different key lengths, wrong sort order)
print(libipld.decode_dag_cbor(bytes.fromhex("a26361616101617802")))
```

They all decode successfully, but they *should* fail, per the spec. (hashberg's dag_cbor, and my own cbrrr, fail as expected)

The first is equivalent to the negative test case you can find here https://github.com/ipld/codec-fixtures/blob/d93d0de6eff0cf2c2efff010c6b2a79aaeb7d100/negative-fixtures/dag-cbor/decode/duplicate-keys.json (I should probably PR the other two test cases there!)

I believe this is an upstream issue with https://github.com/ipld/serde_ipld_dagcbor but I don't know enough rust to make a neat test case for it, so I'm hoping you can help me move this up the chain :)

**Comment by MarshalX** [comment]
yeeeah. i recently read specification of dag-cbor. the order it not strict anymore. but you are right the error must be throw on key duplication. i have todo in the code for it xd 

https://github.com/MarshalX/python-libipld/blob/59cf00003b3002c2c9f098f5223e406470f1f638/src/lib.rs#L82

moreover by current impl allows map keys not only for strings types. i'll change it soon too. and i hope it will give some small perf boost too

**Comment by MarshalX** [comment]
btw i have in my todo list adding of https://github.com/ipld/codec-fixtures/tree/master/fixtures for the tests. i alreade have some from popular json benchmarks https://github.com/MarshalX/python-libipld/tree/main/data

**Comment by DavidBuchanan314** [comment]
I'm pretty sure key ordering is still strict? https://ipld.io/specs/codecs/dag-cbor/spec/#strictness

"The keys in every map must be sorted length-first by the byte representation of the string keys"

**Comment by MarshalX** [comment]
> I'm pretty sure key ordering is still strict? https://ipld.io/specs/codecs/dag-cbor/spec/#strictness
> 
> "The keys in every map must be sorted length-first by the byte representation of the string keys"

ah. i was confused because of different stricness for encoding and decoding. for decoding it's fine to not keep order. thank you!

**Comment by DavidBuchanan314** [comment]
Ah good point, seems like a "non-strict" decoder is allowed to allow them - I think we should be strict by default in the atproto-verse :)

**Comment by MarshalX** [comment]
mb i will add "strict" bool for decoding

**Comment by DavidBuchanan314** [comment]
Hm it looks like `libipld.encode_dag_cbor()` is also non-strict with the order of map keys it emits

```py
import libipld
import dag_cbor

encoded = libipld.encode_dag_cbor({'aaa': 1, 'x': 2}) # b'\xa2caaa\x01ax\x02'
dag_cbor.decode(encoded) # raises dag_cbor.decoding.err.DAGCBORDecodingError
```

**Comment by MarshalX** [comment]
you found all my TODOs xd

**Comment by MarshalX** [comment]
do you check "duplicate key" on encode? fixed in decode https://github.com/MarshalX/python-libipld/pull/16

**Comment by DavidBuchanan314** [comment]
I don't check on encode because (as far as I know) it's impossible for a python dict to have duplicate keys in the first place

**Comment by MarshalX** [comment]
fix for keys order:
- https://github.com/MarshalX/python-libipld/pull/18

**Comment by MarshalX** [comment]
fixed in v1.2.2. published on pypi already. thank you!! btw it worth to rerun benchmarks cuz perf was affected too
